zigzag/classes/stages/SalsaStage.py

```diff
@@ -0,0 +1,151 @@
+
+
+    def run(self):
+        """
+        Set up and start salsa engine, then collect and return the best cost model evaluation
+        """
+
+        logger.info(f"Running SALSA Temporal Mapping Optimizer with {self.number_of_core_allocated} core(s).")
+
+        self.engine = SalsaEngine(accelerator=self.accelerator, 
+                                    layer=self.layer, 
+                                    spatial_mapping=self.spatial_mapping, 
+                                    **self.kwargs)
+
+        # self.best_cme = self.engine.run(self.cme_queue)
+
+        # Get the number of core the user wants to allocate
+        if self.number_of_core_allocated <= multiprocessing.cpu_count():
+            self.number_of_core = self.number_of_core_allocated
+        else:
+            self.number_of_core = multiprocessing.cpu_count()
+
+        # Create processes
+        for core_id in range(0,self.number_of_core):
+            p = multiprocessing.Process(target=self.engine.run, args=(self.cme_queue,))
+            self.worker_list.append(p)
+
+        # Start the processes
+        for core_id in range(0,self.number_of_core):
+            logger.debug(f"Starting SALSA Process #{core_id}.")
+            self.worker_list[core_id].start()
+
+        # For every core we gather the ouput
+        for core_id in range(0,self.number_of_core):
+            cme = self.cme_queue.get()
+            self.compare_stage(cme)
+
+        # Then join them to make sure they all end before continuing the execution
+        for core_id in range(0,self.number_of_core):
+            self.worker_list[core_id].join()
+
+
+        kwargs = self.kwargs.copy()
+        kwargs['accelerator'] = self.accelerator
+        kwargs['layer'] = self.layer
+        kwargs['spatial_mapping'] = self.spatial_mapping
+        kwargs['temporal_mapping'] = self.best_cme.mapping.temporal_mapping
+        sub_stage = self.list_of_callables[0](self.list_of_callables[1:], **kwargs)
+
+        for cme, extra_info in sub_stage.run():
+            yield cme, (self.best_cme.mapping.temporal_mapping, extra_info)
+
+
+    def compare_cme_latency(self, cme):
+        """
+        Compare the latency of the current cost model evaluation with the best latency found so far.
+        Then replace the current best cme if the current cme has a lower latency.
+        """
+
+        if self.best_cme is None:
+            self.best_cme = cme
+        elif cme.latency_total0 < self.best_cme.latency_total0:
+            self.best_cme = cme
+
+
+    def compare_cme_energy(self, cme):
+        """
+        Compare the energy of the current cost model evaluation with the best energy found so far.
+        Then replace the best cme if the current cme has a lower energy.
+        """
+        if self.best_cme is None:
+            self.best_cme = cme
+        elif cme.energy_total < self.best_cme.energy_total:
```

Comment:
Also add that if energies are equal but the new cme's latency is better, we pick the new cme as best cme.